content/architecture/storage-engine/prolly-tree.md

```diff
@@ -2,38 +2,234 @@
+
+When we insert a key, we walk the tree by key to the leaf node where the key belongs. We then edit the chunk, calculate whether to split the chunk or not, and then recalculate the content address. Note, a key insert or delete has a small probability of splitting the chunk. If the rolling hash value when incorporating the new key randomly comes up below our boundary threshold, the chunk is split at the new boundary. Finally, we walk up the tree recalculating each content address up to the root of the tree.
+
+In Dolt's Prolly Tree implementation, chunks are set to be on average 4 kilobytes. This means that you have a base probability of 1/4096 or 0.02% of triggering a chunk boundary shift with a single byte rolling hash window. A single byte window would not be enough data to provide the requisite level of randomness so the rolling hash window is set to 64 bytes, providing a base probability of a chunk split of 64/4096 or 1.6%. As we'll discuss later, the size of the chunk is considered when computing the rolling hash. So, the larger the chunk, the greater the probability the chunk will be split on the addition of a new key. Conversely, for small chunks, the probability of a chunk split is very small. This means that every edit to a table in Dolt is a minimum of 4Kb multiplied by the depth of the tree.
+
+#### At the Beginning
+
+When you insert a key at the beginning of the key space, the tree is modified along the left edge.
+
+![Prolly Tree Beginning Insert](../../.gitbook/assets/prolly-tree-front-insert.png)
+
+#### At the end
+
+When you insert a key at the end of the key space, the tree is modified along the right edge.
+
+![Prolly Tree End Insert](../../.gitbook/assets/prolly-tree-back-insert.png)
+
+#### In the middle
+
+When you insert a key in the middle of the key space, the tree is modified along a spline.
+
+![Prolly Tree Middle Insert](../../.gitbook/assets/prolly-tree-middle-insert.png)
+
+### Delete a key
+
+When you delete a key, the tree is modified under the same rules as an insert.
+
+![Prolly Tree Delete Key](../../.gitbook/assets/prolly-tree-delete.png)
+
+# Properties
+
+## History Independence
+
+A key property of a Prolly tree that enables fast diff and structural sharing is history independence. No matter which order you insert, update, or delete values, the Prolly tree is the same. This is best seen through example.
+
+Consider a map with 4 integer keys, `(1, 2, 3, 4)` pointing at the same values. Before hand we know the combination of keys `(1, 2)` will trigger a chunk boundary. We can know this beforehand because chunk boundaries are based on the size of the chunk and its contents. No matter which order we insert, update, or delete, if we will end up with a tree with two leaf nodes `(1, 2)`, and `(3,4)`. This also happens to be true if the keys have different values. The tree will be the same but the the chunks will have different addresses.
```

Comment:
```suggestion
Consider a map with 4 integer keys, `(1, 2, 3, 4)` pointing at the same values. Before hand we know the combination of keys `(1, 2)` will trigger a chunk boundary. We can know this beforehand because chunk boundaries are based on the size of the chunk and its contents. No matter which order we insert, update, or delete, if we will end up with a tree with two leaf nodes `(1, 2)`, and `(3,4)`. This also happens to be true if the values at these keys are different. The tree will be the same but the the chunks will have different addresses.
```

Not sure if it's worth pointing out this is only true for same-sized values.